zerver/tests/test_messages.py

```diff
@@ -1575,6 +1575,104 @@ def test_message_fail_to_announce(self) -> None:
                                                     "topic": "Test topic"})
         self.assert_json_error(result, "Only organization administrators can send to this stream.")
 
+    def test_successful_message_as_stream_post_policy_restrict_new_members(self) -> None:
+        """
+        Sending a successful message to streams which new members cannot create and post to,
+        by a new realm admin.
+        """
+        user_profile = self.example_user("iago")
+        self.login(user_profile.email)
+
+        do_set_realm_property(user_profile.realm, 'waiting_period_threshold', 10)
+        user_profile.date_joined = timezone_now() - datetime.timedelta(days=9)
+        user_profile.save()
+        self.assertTrue(user_profile.is_new_member)
+        self.assertTrue(user_profile.is_realm_admin)
+
+        stream_name = "Verona"
+        stream = get_stream(stream_name, user_profile.realm)
+        stream.stream_post_policy = Stream.STREAM_POST_POLICY_RESTRICT_NEW_MEMBERS
+        stream.save()
+        result = self.client_post("/json/messages", {"type": "stream",
+                                                     "to": stream_name,
+                                                     "client": "test suite",
+                                                     "content": "Test message",
+                                                     "topic": "Test topic"})
+        self.assert_json_success(result)
+
+        admin_owned_bot = self.create_test_bot(
+            short_name='whatever',
+            user_profile=user_profile,
+        )
+        result = self.api_post(admin_owned_bot.email,
+                               "/api/v1/messages", {"type": "stream",
+                                                    "to": stream_name,
+                                                    "client": "test suite",
+                                                    "content": "Test message",
+                                                    "topic": "Test topic"})
+        self.assert_json_success(result)
+
+        # Cross realm bots should be allowed (through internal_send_message)
+        notification_bot = get_system_bot("notification-bot@zulip.com")
+        internal_send_stream_message(stream.realm, notification_bot, stream,
+                                     'Test topic', 'Test message by notification bot')
+        self.assertEqual(self.get_last_message().content, 'Test message by notification bot')
```

Comment:
I'm not a huge fan of the test setup of having two functions splitting out the success/unsuccessful cases.  It's OK, but I think it's more readable to have a single test function that sets up 4 users in the different states (guest, non-full member, full member, new admin, old admin) and then goes through each configuration and tests which users can send to the streams in that configuration.  Can you think about whether you can clean these tests up using that approach?  I'm not 100% sure it'll be better.

More importantly, I think `self.send_stream_message` might let you write these `client_post()`  lines with a lot less duplication of irrelevant-to-the-test fields like client/content/topic.